File: apps/packages/product-client/src/domain/diagnostics/sequence.ts
import type { ProducerRecordV1, RejectionReasonV1 } from "./contract";
import { DiagnosticsContractErrorV1 } from "./validation-support";
import { parseProducerRecordV1 } from "./validation";

export type SequenceDispositionV1 =
  | "accepted_start"
  | "accepted_terminal"
  | "duplicate"
  | "orphan_terminal";

export interface SequenceDecisionV1 {
  disposition: SequenceDispositionV1;
  requires_gap_or_health_violation: boolean;
}

interface OperationStateV1 {
  start?: ProducerRecordV1;
  terminal?: ProducerRecordV1;
}

export class LifecycleSequenceValidatorV1 {
  private readonly operations = new Map<string, OperationStateV1>();

  apply(input: unknown): SequenceDecisionV1 {
    const record = parseProducerRecordV1(input);
    const lifecycle = record.lifecycle;
    if (lifecycle === undefined) {
      fail("invalid_shape");
    }
    const state = this.operations.get(record.operation_id);
    if (state !== undefined) {
      if (state.terminal !== undefined) {
        if (sameRecord(state.terminal, record)) {
          return decision("duplicate", false);
        }
        fail("conflicting_terminal");
      }
      if (lifecycle.phase === "started") {
        if (state.start !== undefined && sameRecord(state.start, record)) {
          return decision("duplicate", false);
        }
        fail("invalid_shape");
      }
      if (state.start === undefined || !sameOperationIdentity(state.start, record)) {
        fail("invalid_shape");
      }
      state.terminal = record;
      return decision("accepted_terminal", false);
    }

    if (lifecycle.phase === "started") {
      this.operations.set(record.operation_id, { start: record });
      return decision("accepted_start", false);
    }
    this.operations.set(record.operation_id, { terminal: record });
    return decision("orphan_terminal", true);
  }
}

function decision(
  disposition: SequenceDispositionV1,
  requiresGapOrHealthViolation: boolean,
): SequenceDecisionV1 {
  return {
    disposition,
    requires_gap_or_health_violation: requiresGapOrHealthViolation,
  };
}

function sameRecord(left: ProducerRecordV1, right: ProducerRecordV1): boolean {
  return JSON.stringify(left) === JSON.stringify(right);
}

function sameOperationIdentity(left: ProducerRecordV1, right: ProducerRecordV1): boolean {
  return (
    left.operation_id === right.operation_id &&
    left.producer_boot_id === right.producer_boot_id &&
    left.component === right.component &&
    left.source === right.source &&
    left.name === right.name
  );
}

function fail(reason: RejectionReasonV1): never {
  throw new DiagnosticsContractErrorV1(reason);
}

File: server/proliferate/lib/product/observability/sequence.py
from dataclasses import dataclass
from typing import Literal, Never, TypedDict

from proliferate.lib.product.observability.contract import (
    ProducerRecordV1,
    RejectionReasonV1,
)
from proliferate.lib.product.observability.validation import parse_producer_record_v1
from proliferate.lib.product.observability.validation_error import (
    DiagnosticsContractErrorV1,
)

type SequenceDispositionV1 = Literal[
    "accepted_start", "accepted_terminal", "duplicate", "orphan_terminal"
]


class SequenceDecisionV1(TypedDict):
    disposition: SequenceDispositionV1
    requires_gap_or_health_violation: bool


@dataclass
class _OperationStateV1:
    start: ProducerRecordV1 | None = None
    terminal: ProducerRecordV1 | None = None


class LifecycleSequenceValidatorV1:
    def __init__(self) -> None:
        self._operations: dict[str, _OperationStateV1] = {}

    def apply(self, value: object) -> SequenceDecisionV1:
        record = parse_producer_record_v1(value)
        lifecycle = record.get("lifecycle")
        if lifecycle is None:
            _fail("invalid_shape")
        state = self._operations.get(record["operation_id"])
        if state is not None:
            if state.terminal is not None:
                if state.terminal == record:
                    return _decision("duplicate", False)
                _fail("conflicting_terminal")
            if lifecycle["phase"] == "started":
                if state.start == record:
                    return _decision("duplicate", False)
                _fail("invalid_shape")
            if state.start is None or not _same_operation_identity(state.start, record):
                _fail("invalid_shape")
            state.terminal = record
            return _decision("accepted_terminal", False)

        if lifecycle["phase"] == "started":
            self._operations[record["operation_id"]] = _OperationStateV1(start=record)
            return _decision("accepted_start", False)
        self._operations[record["operation_id"]] = _OperationStateV1(terminal=record)
        return _decision("orphan_terminal", True)


def _decision(
    disposition: SequenceDispositionV1, requires_gap_or_health_violation: bool
) -> SequenceDecisionV1:
    return {
        "disposition": disposition,
        "requires_gap_or_health_violation": requires_gap_or_health_violation,
    }


def _same_operation_identity(left: ProducerRecordV1, right: ProducerRecordV1) -> bool:
    return (
        left["operation_id"] == right["operation_id"]
        and left["producer_boot_id"] == right["producer_boot_id"]
        and left["component"] == right["component"]
        and left["source"] == right["source"]
        and left["name"] == right["name"]
    )


def _fail(reason: RejectionReasonV1) -> Never:
    raise DiagnosticsContractErrorV1(reason)
